internal/pkg/pkg.go

```diff
@@ -436,6 +466,117 @@ func Subpackages(rootPath string, matcher SubpackageMatcher, recursive bool) ([]
 	return paths, nil
 }
 
+// SubpackagesWithFs returns a slice of paths to any subpackages of the provided path.
+// The matcher parameter decides if all types of subpackages should be considered,
+// and the recursive parameter determines if only direct subpackages are
+// considered. All returned paths will be relative to the provided rootPath.
+// The top level package is not considered a subpackage. If the provided path
+// doesn't exist, an empty slice will be returned.
+// It operates on the input FileSystem as opposed to FileSystem on disk
+func SubpackagesWithFs(fsys filesys.FileSystem, rootPath string, matcher SubpackageMatcher, recursive bool) ([]string, error) {
```

Comment:
I am not 100% sure how kpt performs rendering of sub-packages. This function seems to discover them. Are they then wrapped in a filesystem that presents just those sub-packages, or does kpt manage that on its own by reading/writing only from those sub-directories? (I assume the latter)

Reply:
It's the latter. During the rendering of the pkg hierarchy, its read-only. Write happens at the end at the end of the `Render` call at the root pkg.